### PR TITLE
Add amqp operations

### PR DESCRIPTION
**Pull request by lcc**
This patch adds all operations used in the configuration progress of a Thing with the KNoT Cloud. As the patch name implies these operations were implemented using the AMQP protocol. 

Depends on #11

### CODE REVIEW

File: app/src/main/java/com/cesar/knot_thing_lib_android/MainActivity.kt [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+import com.cesar.knot_sdk.KNoTAMQPFactory
+import com.cesar.knot_sdk.KNoTMessager
+import com.cesar.knot_sdk.knot_messages.KNoTThingRegister
+import kotlinx.android.synthetic.main.activity_main.register_button
+import org.jetbrains.anko.doAsync
+
+class MainActivity : AppCompatActivity() {
+    val HOSTNAME = "192.168.31.61"
+    val PORT_NUMBER = 5672
+    val USERNAME = "lucas"
```
Comment (alla3) [inline app/src/main/java/com/cesar/knot_thing_lib_android/MainActivity.kt]: Avoid using hard-coded values. You can use generic credentials or a configuration file instead of using your credentials.

File: app/src/main/java/com/cesar/knot_thing_lib_android/MainActivity.kt [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    val PASSWORD = "lucas"
+    val THING_ID = "a74151d19de59cd3"
+    val THING_NAME = "pocophone-lucas"
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_main)
+        val kNoTAMQP = KNoTAMQP(USERNAME, PASSWORD, HOSTNAME, PORT_NUMBER)
```
Comment (alla3) [inline app/src/main/java/com/cesar/knot_thing_lib_android/MainActivity.kt]: I would use kNoT in lowercase.

File: app/src/main/AndroidManifest.xml [hunk 3]
```diff
@@ -10,7 +10,7 @@
         android:roundIcon="@mipmap/ic_launcher_round"
         android:supportsRtl="true"
         android:theme="@style/AppTheme">
-        <activity android:name=".MainActivity">
```
Comment (ehammo) [inline app/src/main/AndroidManifest.xml]: Is this change necessary?

File: app/src/main/java/com/cesar/knot_thing_lib_android/MainActivity.kt [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+        lateinit var kNoTMessager : KNoTMessager
+        val setKNoTMessager = { kNoTMessagerAux : KNoTMessager -> kNoTMessager = kNoTMessagerAux }
```
Comment (ehammo) [inline app/src/main/java/com/cesar/knot_thing_lib_android/MainActivity.kt]: Maybe use corotines here?

File: app/src/main/res/layout/activity_main.xml [hunk 5]
```diff
@@ -0,0 +1,13 @@
+<?xml version="1.0" encoding="utf-8"?>
+<LinearLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    android:orientation="vertical" android:layout_width="match_parent"
+    android:layout_height="match_parent">
+
+    <Button
+        android:id="@+id/register_button"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:text="send-register-message"
```
Comment (ehammo) [inline app/src/main/res/layout/activity_main.xml]: Please move to res

File: knot-thing/src/main/java/com/cesar/knot_sdk/KNoTAMQP.kt [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+    init {
+        factory.host = hostname
+        factory.username = username
+        factory.password = password
+        factory.port = port
+    }
+
+    /**
+     * Starts the connection with the rabbitMQ server. Uses the parameters
+     * defined in factory. Should not be executed in the main thread as it
```
Comment (ehammo) [inline knot-thing/src/main/java/com/cesar/knot_sdk/KNoTAMQP.kt:58]: then shouldn't it be a suspended function? Or an async function?

File: knot-thing/src/main/java/com/cesar/knot_sdk/knot_messages/KNoTThingRegister.kt [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+/**
+ * This class is an abstraction for the registration message. Its property names
+ * are already mapped to the expected values in the message json, which means
+ * it's really easy to convert these values to the expected json message.
+ * @property id a unique identifier for the Thing, be careful to not generate a
+ * duplicate name, as the KNoT-Protocol does not define what happens when
+ * receiving a duplicate id.
+ * @property name a string that represents this device in some manner.
+ */
+data class KNoTThingRegister(
```
Comment (ehammo) [inline knot-thing/src/main/java/com/cesar/knot_sdk/knot_messages/KNoTThingRegister.kt]: Is not clear to me that this class is an abstraction for the registration message by looking only to the name.

File: app/src/androidTest/java/com/cesar/knot_thing_lib_android/ExampleInstrumentedTest.kt [hunk 8]
```diff
@@ -0,0 +1,24 @@
+package com.cesar.knot_thing_lib_android
```
Comment (ehammo) [inline app/src/androidTest/java/com/cesar/knot_thing_lib_android/ExampleInstrumentedTest.kt]: please remove this class if you're not testing anything.

File: app/src/test/java/com/cesar/knot_thing_lib_android/ExampleUnitTest.kt [hunk 9]
```diff
@@ -0,0 +1,17 @@
+package com.cesar.knot_thing_lib_android
```
Comment (ehammo) [inline app/src/test/java/com/cesar/knot_thing_lib_android/ExampleUnitTest.kt]: please remove this class if you are not testing anything

File: knot-thing/src/main/java/com/cesar/knot_sdk/KNoTAMQPFactory.kt [hunk 10]
```diff
@@ -44,6 +44,13 @@ class KNoTAMQPFactory() {
                 kNoTAMQP.EXCHANGE_NAME_FOG,
                 kNoTAMQP.BINDING_KEY_DATA_PUBLISH)
 
+            kNoTAMQP.bindQueue(kNoTAMQP.QUEUE_NAME_FOG_OUT,
```
Comment (ehammo) [inline knot-thing/src/main/java/com/cesar/knot_sdk/KNoTAMQPFactory.kt]: Please fix your indentation

File: app/src/main/res/layout/activity_main.xml [hunk 11]
```diff
@@ -31,4 +31,11 @@
         android:text="send-schema-update-message"
         android:layout_gravity="center" />
 
+    <Button
+        android:id="@+id/data_publish_button"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:text="send-data-message"
```
Comment (ehammo) [inline app/src/main/res/layout/activity_main.xml]: Extract string to res

File: knot-thing/src/main/java/com/cesar/knot_sdk/KNoTAMQPFactory.kt [hunk 12]
```diff
@@ -40,6 +40,10 @@ class KNoTAMQPFactory() {
                 kNoTAMQP.EXCHANGE_NAME_FOG,
                 kNoTAMQP.BINDING_KEY_SCHEMA_UPDATE)
 
+            kNoTAMQP.bindQueue(kNoTAMQP.QUEUE_NAME_FOG_IN,
```
Comment (ehammo) [inline knot-thing/src/main/java/com/cesar/knot_sdk/KNoTAMQPFactory.kt]: Please fix indentation

File: knot-thing/src/main/java/com/cesar/knot_sdk/KNoTTypes.kt [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+    const val KNOT_UNIT_ENERGY_KCAL             =       0x06
+
+// definition of value type
+    const val KNOT_VALUE_NULL                   =       0x00
+    const val KNOT_VALUE_TYPE_INT               =       0x01
+    const val KNOT_VALUE_TYPE_FLOAT             =       0x02
+    const val KNOT_VALUE_TYPE_BOOL              =       0x03
+    const val KNOT_VALUE_TYPE_RAW               =       0x04
+    const val KNOT_VALUE_TYPE_MIN               =       KNOT_VALUE_TYPE_INT
+    const val KNOT_VALUE_TYPE_MAX               =       KNOT_VALUE_TYPE_RAW + 1
```
Comment (ehammo) [inline knot-thing/src/main/java/com/cesar/knot_sdk/KNoTTypes.kt:129]: Why the MAX is not RAW?

**Comment by ehammo** [comment]
The commits are in a order that does not make a lot of sense.

**Comment by ehammo** [comment]
This PR should depend on another PR?

File: app/src/main/AndroidManifest.xml (hunk 3 above)
Comment (lcc) [reply app/src/main/AndroidManifest.xml]: Done.

File: app/src/main/java/com/cesar/knot_thing_lib_android/MainActivity.kt (hunk 4 above)
Comment (lcc) [reply app/src/main/java/com/cesar/knot_thing_lib_android/MainActivity.kt]: I changed this excerpt but didn't deem necessary to use coroutines, could you please check it again?

File: app/src/main/res/layout/activity_main.xml (hunk 5 above)
Comment (lcc) [reply app/src/main/res/layout/activity_main.xml]: Done.

File: knot-thing/src/main/java/com/cesar/knot_sdk/KNoTAMQP.kt (hunk 6 above)
Comment (lcc) [reply knot-thing/src/main/java/com/cesar/knot_sdk/KNoTAMQP.kt:58]: The class KNoTAMQPFactory takes care of creating a valid AMQP instance, that includes doing network operations in an async manner. For that I used "doAsyc" from the anko library. This method was preferred over putting the "doAsync" inside the KNoTAMQP class as certain operations are required to be ran sequentially (e.g.: a connection should be started before the createChannel method is called) which would not be guaranteed if each method ran in a separate thread.

Comment (lcc) [reply knot-thing/src/main/java/com/cesar/knot_sdk/KNoTAMQP.kt:58]: Given this,  do you have any recommendations for a better way to solve this issue?

File: knot-thing/src/main/java/com/cesar/knot_sdk/knot_messages/KNoTThingRegister.kt (hunk 7 above)
Comment (lcc) [reply knot-thing/src/main/java/com/cesar/knot_sdk/knot_messages/KNoTThingRegister.kt]: Done.

File: app/src/androidTest/java/com/cesar/knot_thing_lib_android/ExampleInstrumentedTest.kt (hunk 8 above)
Comment (lcc) [reply app/src/androidTest/java/com/cesar/knot_thing_lib_android/ExampleInstrumentedTest.kt]: Done.

File: app/src/test/java/com/cesar/knot_thing_lib_android/ExampleUnitTest.kt (hunk 9 above)
Comment (lcc) [reply app/src/test/java/com/cesar/knot_thing_lib_android/ExampleUnitTest.kt]: Done.

File: knot-thing/src/main/java/com/cesar/knot_sdk/KNoTAMQPFactory.kt (hunk 10 above)
Comment (lcc) [reply knot-thing/src/main/java/com/cesar/knot_sdk/KNoTAMQPFactory.kt]: Done.

File: app/src/main/res/layout/activity_main.xml (hunk 11 above)
Comment (lcc) [reply app/src/main/res/layout/activity_main.xml]: Done.

File: knot-thing/src/main/java/com/cesar/knot_sdk/KNoTAMQPFactory.kt (hunk 12 above)
Comment (lcc) [reply knot-thing/src/main/java/com/cesar/knot_sdk/KNoTAMQPFactory.kt]: Done.

File: knot-thing/src/main/java/com/cesar/knot_sdk/KNoTTypes.kt (hunk 13 above)
Comment (lcc) [reply knot-thing/src/main/java/com/cesar/knot_sdk/KNoTTypes.kt:129]: Since there is a C implementation for this same class, I assume they are checking whether a value is lesser than another, rather than lesser than equal.

File: knot-thing/src/main/java/com/cesar/knot_sdk/KNoTAMQPFactory.kt [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+            kNoTAMQP.createQueue(kNoTAMQP.QUEUE_NAME_FOG_OUT)
+
+            kNoTAMQP.bindQueue(
+                kNoTAMQP.QUEUE_NAME_FOG_IN,
+                kNoTAMQP.EXCHANGE_NAME_CLOUD,
+                kNoTAMQP.BINDING_KEY_REGISTER
+            )
+        }
+    }
+}
```
Comment (alla3) [inline knot-thing/src/main/java/com/cesar/knot_sdk/KNoTAMQPFactory.kt]: Add a newline at the end of the file.

File: knot-thing/src/main/java/com/cesar/knot_sdk/KNoTTypes.kt [hunk 15]
```diff
@@ -0,0 +1,129 @@
+package com.cesar.knot_sdk
+
+/**
+ * This is a collection of all supported KNoT types, units and  values.
```
Comment (alla3) [inline knot-thing/src/main/java/com/cesar/knot_sdk/KNoTTypes.kt]: Add in this file and at the commit message, this file is based on KNoT Protocol Repository with the link and the commit hash you use as a reference.

File: knot-thing/src/main/java/com/cesar/knot_sdk/KNoTMessager.kt [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+/**
+ * This interface represents all operations that are available with the KNoT
+ * protocol.
+ * Classes that implement this interface can use any kind of technology to send
+ * these messages (AMQP, WebSockets, etc), but should support these operations.
+ */
+interface KNoTMessager {
+
+    /**
+     * Register a KNoT Thing in a KNoT Cloud.
```
Comment (netoax) [inline knot-thing/src/main/java/com/cesar/knot_sdk/KNoTMessager.kt]: Remember we shouldn't care about the cloud solution or provider when mentioning an operation on it, so I think you can change this comment to "Register a KNoT Thing in the cloud" or similar.

File: knot-thing/src/main/java/com/cesar/knot_sdk/KNoTAMQP.kt [hunk 17]
```diff
@@ -0,0 +1,139 @@
...
+
+    /**
+     * Creates a durable, non-exclusive and non self-deleting queue without
+     * extra parameters with the given name.
+     * @param queueName the name of the queue, must be either fogIn or fogOut.
+     */
+    fun createQueue(queueName: String) {
+        channel.queueDeclare(
+            queueName,
+            true,
```
Comment (alla3) [inline knot-thing/src/main/java/com/cesar/knot_sdk/KNoTAMQP.kt]: What are these parameters? Maybe you could use declared values for that.

File: knot-thing/src/main/java/com/cesar/knot_sdk/KNoTAMQPController.kt [hunk 18]
```diff
@@ -0,0 +1,55 @@
...
+import com.cesar.knot_sdk.knot_messages.KNoTMessageUnregister
+import com.cesar.knot_sdk.knot_messages.KNoTMessageAuth
+import com.cesar.knot_sdk.knot_messages.KNoTMessageUpdateSchema
+import com.cesar.knot_sdk.knot_messages.KNoTMessageUpdateData
+import com.google.gson.Gson
+
+/**
+ * This class implements the KNoT-Protocol operations for AMQP.
+ */
+class KNoTAMQPController(val kNoTAMQP: KNoTAMQP) : KNoTMessager {
```
Comment (netoax) [inline knot-thing/src/main/java/com/cesar/knot_sdk/KNoTAMQPController.kt]: I see the parameter `kNoTAMQP ` as being similar to the overridden methods parameters below, i.e. it must start with `knot` instead of `kNoT` in order to keep following a pattern.

File: knot-thing/src/main/java/com/cesar/knot_sdk/KNoTAMQP.kt [hunk 19]
```diff
@@ -0,0 +1,139 @@
...
+        )
+    }
+
+    /**
+     * This method is responsible for releasing the resources used by the AMQP
+     * communication. This method should only be called when ending a connection
+     * (when there is no intention of publishing/consuming data). Ideally this
+     * method should only be called when the application is being destroyed.
+     */
+    fun disconnect() {
```
Comment (alla3) [inline knot-thing/src/main/java/com/cesar/knot_sdk/KNoTAMQP.kt:179]: You are not using this function. I can see when it should be used but is a better practice to use every function you add.

File: app/src/main/java/com/cesar/knot_thing_lib_android/MainActivity.kt [hunk 20]
```diff
@@ -49,7 +52,18 @@ class MainActivity : AppCompatActivity() {
             )
         )
 
-        val kNoTThingUpdateSchema = KNoTMessageUpdateSchema(THING_ID, knotThingSchema)
+        val knotThingUpdateSchema = KNoTMessageUpdateSchema(THING_ID, knotThingSchema)
```
Comment (alla3) [inline app/src/main/java/com/cesar/knot_thing_lib_android/MainActivity.kt:55]: This change should not be on commit 39798759c1b3214c34c101fa6ef9b7c0bc49380d.

File: app/src/main/java/com/cesar/knot_thing_lib_android/MainActivity.kt [hunk 21]
```diff
@@ -64,7 +78,11 @@ class MainActivity : AppCompatActivity() {
         }
 
         schema_update_button.setOnClickListener {
-            doAsync { kNoTMessager.updateSchema(kNoTThingUpdateSchema) }
+            doAsync { kNoTMessager.updateSchema(knotThingUpdateSchema) }
```
Comment (alla3) [inline app/src/main/java/com/cesar/knot_thing_lib_android/MainActivity.kt]: This change should not be on commit 39798759c1b3214c34c101fa6ef9b7c0bc49380d.

**Review by alla3 — CHANGES_REQUESTED** [review]

File: knot-thing/src/main/java/com/cesar/knot_sdk/KNoTAMQPFactory.kt [hunk 22]
```diff
@@ -0,0 +1,69 @@
...
+ * responsible solely to ease the creation of KNoTAMQP instances.
+ */
+class KNoTAMQPFactory() {
+
+    /**
+     * Configures the KNoT-AMQP instance.
+     * @param kNoTAMQP a non-configured instance of the KNoTAMQP class.
+     * @param callback a callback to return after the operations are done.
+     */
+    fun getKNoTProtocolMessager(
```
Comment (netoax) [inline knot-thing/src/main/java/com/cesar/knot_sdk/KNoTAMQPFactory.kt]: I'd change this name to `create()` or `getKNoTAMQP()`.

Despite of being JavaScript, [here](https://github.com/CESARBR/knot-fog-connector/blob/master/src/network/AMQPConnectionFactory.js#L8) is one example of using `create()` in a factory and [here](https://github.com/CESARBR/knot-fog-connector/blob/master/src/index.js#L22) is one example of accessing it.

File: knot-thing/src/main/java/com/cesar/knot_sdk/KNoTAMQPFactory.kt [hunk 23]
```diff
@@ -0,0 +1,69 @@
+package com.cesar.knot_sdk
+
+import org.jetbrains.anko.doAsync
+
+/**
+ * The creation of the KNoTAMQP class depends on a series of complicated
+ * operations, some of which involve network operations. This class is
+ * responsible solely to ease the creation of KNoTAMQP instances.
+ */
+class KNoTAMQPFactory() {
```
Comment (netoax) [inline knot-thing/src/main/java/com/cesar/knot_sdk/KNoTAMQPFactory.kt]: As you explained a few minutes ago, these parentheses could be removed here.

File: knot-thing/src/main/java/com/cesar/knot_sdk/KNoTTypes.kt (hunk 15 above)
Comment (netoax) [inline knot-thing/src/main/java/com/cesar/knot_sdk/KNoTTypes.kt]: "This is a collection of all supported KNoT types, units and **values."**

**Review by netoax — APPROVED** [review]

File: app/src/main/java/com/cesar/knot_thing_lib_android/MainActivity.kt (hunk 1 above)
Comment (netoax) [reply app/src/main/java/com/cesar/knot_thing_lib_android/MainActivity.kt]: +1

File: app/src/main/java/com/cesar/knot_thing_lib_android/MainActivity.kt (hunk 2 above)
Comment (netoax) [reply app/src/main/java/com/cesar/knot_thing_lib_android/MainActivity.kt]: +1

File: knot-thing/src/main/java/com/cesar/knot_sdk/KNoTTypes.kt (hunk 15 above)
Comment (netoax) [reply knot-thing/src/main/java/com/cesar/knot_sdk/KNoTTypes.kt]: +1

File: knot-thing/src/main/java/com/cesar/knot_sdk/KNoTAMQP.kt [hunk 24]
```diff
@@ -0,0 +1,180 @@
...
+            exchange,
+            routingKey,
+            messageProperties,
+            messageBodyBytes
+        )
+    }
+
+    /**
+     * Creates a consumer that listens to an AQMP queue. This is used to listen to the messages
+     * being sent by the cloud to the Thing. The queue name should be one that
```
Comment (netoax) [inline knot-thing/src/main/java/com/cesar/knot_sdk/KNoTAMQP.kt]: **"The queue name should be one that .... ?"**

File: knot-thing/src/main/java/com/cesar/knot_sdk/KNoTAMQP.kt [hunk 25]
```diff
@@ -0,0 +1,180 @@
...
+        channel.basicPublish(
+            exchange,
+            routingKey,
+            messageProperties,
+            messageBodyBytes
+        )
+    }
+
+    /**
+     * Creates a consumer that listens to an AQMP queue. This is used to listen to the messages
```
Comment (netoax) [inline knot-thing/src/main/java/com/cesar/knot_sdk/KNoTAMQP.kt]: **AMQP**

File: knot-thing/src/main/java/com/cesar/knot_sdk/KNoTTypes.kt [hunk 26]
```diff
@@ -0,0 +1,129 @@
...
+ */
+object KNoTTypes {
+// definition of invalid sensor id
```
Comment (netoax) [inline knot-thing/src/main/java/com/cesar/knot_sdk/KNoTTypes.kt]: Please, indent the inline comments.

File: knot-thing/src/main/java/com/cesar/knot_sdk/knot_messages/KNoTMessageDataItem.kt [hunk 27]
```diff
@@ -0,0 +1,10 @@
+package com.cesar.knot_sdk.knot_messages
+
+/**
+ * This is the class that is serialized for one data value, from one sensor.
+ * @property id the Thing unique identifier
+ * @property data the data batch that is to be deployed
```
Comment (netoax) [inline knot-thing/src/main/java/com/cesar/knot_sdk/knot_messages/KNoTMessageDataItem.kt]: I got what you mean here but I'd change from `deployed` to `sent`. At least for me, it becomes clearer.

Comment (netoax) [reply knot-thing/src/main/java/com/cesar/knot_sdk/knot_messages/KNoTMessageDataItem.kt]: What do you think about it?

File: knot-thing/src/main/java/com/cesar/knot_sdk/knot_messages/KNoTMessageRegister.kt [hunk 28]
```diff
@@ -0,0 +1,15 @@
+package com.cesar.knot_sdk.knot_messages
+
+/**
+ * This class is an abstraction for the registration message. Its property names
+ * are already mapped to the expected values in the message json, which means
```
Comment (netoax) [inline knot-thing/src/main/java/com/cesar/knot_sdk/knot_messages/KNoTMessageRegister.kt]: `JSON`instead of `json`.

**Review by netoax — CHANGES_REQUESTED** [review]

**Review by alla3 — APPROVED** [review]

File: knot-thing/src/main/java/com/cesar/knot_sdk/KNoTAMQP.kt [hunk 29]
```diff
@@ -0,0 +1,184 @@
...
+                @Throws(IOException::class)
+                override fun handleDelivery(
+                    consumerTag: String,
+                    envelope: Envelope,
+                    properties: AMQP.BasicProperties,
+                    body: ByteArray
+                ) {
+                    val deliveryTag = envelope.getDeliveryTag()
+                    val bodyJson = String(body)
+                    val kNoTThingRegistered = Gson().fromJson(
```
Comment (joaoaneto) [inline knot-thing/src/main/java/com/cesar/knot_sdk/KNoTAMQP.kt]: `knotThingRegistered`*

File: knot-thing/src/main/java/com/cesar/knot_sdk/knot_messages/KNoTMessageRegister.kt (hunk 28 above)
Comment (netoax) [reply knot-thing/src/main/java/com/cesar/knot_sdk/knot_messages/KNoTMessageRegister.kt]: Did you change it?

File: knot-thing/src/main/java/com/cesar/knot_sdk/KNoTAMQP.kt (hunk 29 above)
Comment (netoax) [inline knot-thing/src/main/java/com/cesar/knot_sdk/KNoTAMQP.kt]: `knotThingRegistered`*

**Review by netoax — CHANGES_REQUESTED** [review]
